lib/astc-encoder/Source/cmake_core.cmake

```diff
@@ -111,40 +111,52 @@ macro(astcenc_set_properties NAME)
+            # like clang-cl won't work as desired.
+            # Since clang-cl is also capable of accepting gnu style
+            # arguments a build will fail when -Werror is specified.
+
             # MSVC compiler defines
-            $<$<CXX_COMPILER_ID:MSVC>:/EHsc>
-            $<$<CXX_COMPILER_ID:MSVC>:/fp:strict>
-            $<$<CXX_COMPILER_ID:MSVC>:/wd4324>
+            $<$<BOOL:${MSVC}>:/EHsc>
+            $<$<BOOL:${MSVC}>:/fp:strict>
+            $<$<BOOL:${MSVC}>:/wd4324>
 
             # G++ and Clang++ compiler defines
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wall>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wextra>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wpedantic>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Werror>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wshadow>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wdouble-promotion>
+            $<$<NOT:${MSVC}>:-Wall>
+            $<$<NOT:${MSVC}>:-Wextra>
+            $<$<NOT:${MSVC}>:-Wpedantic>
+            $<$<NOT:${MSVC}>:-Werror>
+            $<$<NOT:${MSVC}>:-Wshadow>
+            $<$<NOT:${MSVC}>:-Wdouble-promotion>
 
             # Hide noise thrown up by Clang 10 and clang-cl
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-unknown-warning-option>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-c++98-compat-pedantic>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-c++98-c++11-compat-pedantic>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-float-equal>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-deprecated-declarations>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-atomic-implicit-seq-cst>
+            $<$<NOT:${MSVC}>:-Wno-unknown-warning-option>
+            $<$<NOT:${MSVC}>:-Wno-c++98-compat-pedantic>
+            $<$<NOT:${MSVC}>:-Wno-c++98-c++11-compat-pedantic>
+            $<$<NOT:${MSVC}>:-Wno-float-equal>
+            $<$<NOT:${MSVC}>:-Wno-deprecated-declarations>
+            $<$<NOT:${MSVC}>:-Wno-atomic-implicit-seq-cst>
 
             # Clang 10 also throws up warnings we need to investigate (ours)
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-cast-align>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-sign-conversion>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-implicit-int-conversion>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-shift-sign-overflow>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-format-nonliteral>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-reserved-identifier>
-            $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wno-cast-function-type>
+            $<$<NOT:${MSVC}>:-Wno-cast-align>
+            $<$<NOT:${MSVC}>:-Wno-sign-conversion>
+            $<$<NOT:${MSVC}>:-Wno-implicit-int-conversion>
+            $<$<NOT:${MSVC}>:-Wno-shift-sign-overflow>
+            $<$<NOT:${MSVC}>:-Wno-format-nonliteral>
+            $<$<NOT:${MSVC}>:-Wno-reserved-identifier>
+            $<$<NOT:${MSVC}>:-Wno-cast-function-type>
```

Comment:
The CXX_COMPILER_ID is a standard syntax for CMake generators, so I'm surprised this doesn't work. What version of CMake are you trying to use?

Reply:
This is a generator expression so this is evaluated later than a regular `if ()` statement. Is the astc-encoder upstream hosted somewhere I can access? I should try building that standalone to ensure that I'm not going crazy

Reply:
It's at https://github.com/ARM-software/astc-encoder.

Reply:
CXX_COMPILER_ID does work as intended. I'm using cmake 3.23.2

The problem is that clang-cl reports "Clang" when it's an msvc-style compiler. So without my change it would apply the gnu-style flags and not apply any of the msvc-style flags for clang-cl. Now clang-cl can also parse and accept gnu-style flags so it didn't outright fail, but -Werror does cause the build to die. 

The cmake var `MSVC` is, AFAIK, the only way to determine if the compiler is msvc-style and it's not usable on its own with generator expressions.

Reply:
I just pulled the upstream astc-encoder and yeah, it's borked. I can see how this happened. Even if the authors are using VS2022 I doubt they're regularly compiling for Windows arm64 which is the only case I can imagine where you'd have the msvc compiler hitting the neon code path. 

You can reproduce this by just running:
* `cmake -B build-msvc-arm -G "Visual Studio 17 2022" -Aarm64 -DISA_NEON=ON`
* `cmake --build build-msvc-arm`

This is a problem with how msvc handles intrinsics. I knew this was the case for SIMD but I didn't know it was also a problem with NEON. What clang and gcc do is for types like `uint32x4_t` the compiler will just supply operators for doing simple accesses. Helpful when you're in C and you can't overload operators. But what MSVC does is it implements `uint32x4_t` as `__n128` which is actually a union. So `mask.m[3]` becomes `mask.m.n128_u32[3]`. 

I have a fix that I can offer upstream so... the journey continues

Reply:
> The problem is that clang-cl reports "Clang" when it's an msvc-style compiler.

I just found [this](https://gitlab.kitware.com/cmake/cmake/-/issues/22668) which says to use `ClangCl ` for this situation.

We have a similar problem between standard clang and Apple clang. In cases where it is a general clang thing we use  `MATCHES "Clang" or generator expression equivalent. If it is clang specific we use `STREQUAL "AppleClang"` or `$<CXX_COMPILER_ID:AppleClang>`.

I think you can do `$<CXX_COMPILER_ID:MSVC,ClangCl>`. One issue with this is that our Emscripten builds still have to use CMake 3.16 and I doubt `ClangCl` is supported there.

Reply:
Whoops. That change has not landed yet.

Reply:
Why does the x64 builds with clang-cl work? Is it that all these compiler options you are modifying relate only to an arm64/Neon build?

Reply:
Yeah I believe that's the case. This all only has to do with the neon code

Reply:
I have just looked at the actual build output and seen there are a large number of compile warnings. All are normally disabled either via pragmas or  CMake `if` statements all based on compiler id and version number. Looks like these need to be updated for clang-cl. For example in `lib/etcdec.cxx` is 

```
#if defined(_MSC_VER)
  #pragma warning(push)
  #pragma warning(disable: 4100 4244 )
#elif __clang__
  #pragma clang diagnostic push
  #pragma clang diagnostic ignored "-Wunused-parameter"
  #if __clang_major__ > 13 || (__clang_major__ == 13 && __clang_minor__ >= 1)
    #pragma clang diagnostic ignored "-Wunused-but-set-variable"
  #endif
#elif __GNUC__
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunused-parameter"
#pragma GCC diagnostic ignored "-Wunused-but-set-variable"
#pragma GCC diagnostic ignored "-Wunused-result" // For the freads.
#endif
```

Others are `CMakeLists.txt`.

Please update these as well.

Reply:
The issue in that case isn't that __clang__ isn't defined it's that _MSC_VER *is* with clang-cl

Reply:
After fiddling with the upstream build of the astc encoder I also have a change that just fixes the warning that clang-cl was complaining about and it compiles cleanly. That may be a better solution